system/jlib/jmetrics.cpp

```diff
@@ -0,0 +1,219 @@
+}
+
+
+MetricsReporter::~MetricsReporter()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        delete sinkIt.second.pSink;
+    }
+}
+
+
+void MetricsReporter::init(IPropertyTree *pMetricsTree)
+{
+    Owned<IPropertyTree> pSinkTree = pMetricsTree->getPropTree("sinks");
+    auto sinkElementsIt = pSinkTree->getElements("sink");
+    initializeSinks(sinkElementsIt);
+}
+
+
+void MetricsReporter::addMetric(const std::shared_ptr<IMetric> &pMetric)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(pMetric->queryName());
+    if (it == metrics.end())
+    {
+        metrics.insert({pMetric->queryName(), pMetric});
+    }
+    // todo - need to decide how to handle duplicate named metrics
+}
+
+
+void MetricsReporter::removeMetric(const std::string &metricName)
+{
+    std::unique_lock<std::mutex> lock(metricVectorMutex);
+    auto it = metrics.find(metricName);
+    if (it != metrics.end())
+    {
+        metrics.erase(it);
+    }
+}
+
+
+void MetricsReporter::startCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->startCollection(this);
+    }
+}
+
+
+void MetricsReporter::stopCollecting()
+{
+    for (auto const &sinkIt : sinks)
+    {
+        sinkIt.second.pSink->stopCollection();
+    }
+}
+
+
+std::vector<std::shared_ptr<IMetric>> MetricsReporter::queryMetricsForReport(const std::string &sinkName)
+{
+    std::vector<std::shared_ptr<IMetric>> reportMetrics;
+
+    auto it = sinks.find(sinkName);
+    if (it != sinks.end())
+    {
+        //
+        // Lock the list of metrics while it's in use
+        std::unique_lock<std::mutex> lock(metricVectorMutex);   // no one else can mess with it for a bit
+        auto metricIt=metrics.begin();
+        while (metricIt != metrics.end())
+        {
+            auto pMetric = metricIt->second.lock();
+            if (pMetric)
+            {
+                // This is where the metric would be compared against the list of metrics to be reported
+                // by the sink (probably a regex). This allows limiting the metrics reported to the sink.
+                // for now, only the default is supported which is reporting all metrics.
+                reportMetrics.emplace_back(pMetric);
+                ++metricIt;
+            }
+            else
+            {
+                metricIt = metrics.erase(metricIt);
+            }
+        }
+    }
+
+    return reportMetrics;  // All metrics
+}
+
+
+
+bool MetricsReporter::initializeSinks(IPropertyTreeIterator *pSinkIt)
+{
+    bool rc = true;
+    for (pSinkIt->first(); pSinkIt->isValid() && rc; pSinkIt->next())
+    {
+        SinkInfo *pSinkInfo;
+        Owned<IPropertyTree> pSinkTree = &pSinkIt->get(); // pSinkIt->query();
+
+        StringBuffer cfgSinkType, cfgSinkName;
+        pSinkTree->getProp("@type", cfgSinkType);  // this one is required
```

Comment:
queryProp() is generally more efficient - preferred if the IPropertyTree will definitely remain alive.